### ISSUE TITLE
Add playlist song functionality

### ISSUE THREAD

**Issue by diraimondo** [thebigmunch/google-music#1]
Hi, your new project is interesting. I'm wondering if it is already possibile to edit a playlist adding some uploaded songs. I can't spot anything similar in API. If missing, please consider to add it.

Such feature could be interesting to add to gms too.

Thanks.

**Comment by thebigmunch** [comment]
I plan on adding all the functionality possible. The rest of the playlist stuff is next on my TODO list after releasing the next major version with the breaking changes already committed. Hopefully, the playlist functionality will be implemented in the next week. I was just too busy the past week or so to get to it like I wanted.

**Comment by thebigmunch** [comment]
PS Before I had started writing this 2 years ago (I never released it mainly due to anxiety issues), I had added playlist functionality into my gmusicapi-wrapper and began adding it into my gmusicapi-wrapper.

So, I will indeed be adding that into google-music-scripts, almost certainly before the end of the year.

And nice to see another notable face from the gmusicapi community.

**Comment by thebigmunch** [comment]
Also, as a preview, I believe the APIs I have in mind for adding and moving songs in playlists will include doing so by position or by specifying another playlist song it should come before or after.

**Comment by thebigmunch** [comment]
Ugh, Google sucks. I knew moving multiple items with a call per item was flaky. But, I can't get a single batch call with multiple items to work for playlist entries unless it's just adding items to the end of a playlist. It's possible that it would work if the entries never had to reference each other for position, but I haven't thoroughly tested that.

Unfortunately, it doesn't seem that I'm doing anything wrong. When I was testing this with my phone, the batch call would be identical to mine. But it would also fail to move all the items properly. So, I dunno if this is something broken on Google's end just right now or how long it's been that way. And I have no way to know if it will be fixed.

TLDR: Single entry additions/moves are working just fine. Multiple entry additions/moves are mostly broken and useless at this time unless I go over things for like the 5th time and have a revelation. I'll have to check back on it in a few days.

**Comment by thebigmunch** [comment]
After some further testing, multiple mutations in one call fail if they depend on one of the other mutations in the call. Don't know why Google even accepts calls like this and return success. This puts a big hamper on the kinds of things I wanted to be possible. But I will try to get out what I can soon.

**Comment by thebigmunch** [comment]
**GOOD NEWS**

I've switched around the implementation and found what I believe to be a reliable method for being able to add/move multiple songs anywhere in a playlist. I'm not sure when I'll throw up a feature branch with some things to try/test. If it doesn't happen before the weekend, it won't be until after.

**Comment by diraimondo** [comment]
Thank you for your efforts!

**Comment by thebigmunch** [comment]
I've pushed  a [feature branch](https://github.com/thebigmunch/google-music/tree/playlist-songs) with a bunch of new methods relating to playlist songs. Besides anyone else giving it a test drive to make sure I didn't screw anything up too badly, I'm particularly interested in the ``position`` parameter. Right now, I have it so ``position`` is akin to the position in the mobile app (starting at 1) rather than Python list indexes (starting at 0). I was super on board with that until right at the very end of testing I found myself screwing up and putting a list index a couple of times. I'd like to hear opinions on that.

**Comment by thebigmunch** [comment]
Almost forgot, if you want to test it out, you'll need to install [this branch](https://github.com/thebigmunch/google-music-proto/tree/playlist-entries-batch) of google-music-proto first.

``pip install -U git+https://github.com/thebigmunch/google-music-proto@playlist-entries-batch git+https://github.com/thebigmunch/google-music@playlist-songs``

**Comment by thebigmunch** [comment]
These still aren't as reliable as I'd like them. See #3 for more info/discussion.
